### DISCUSSION TITLE
Performance issues

### DISCUSSION THREAD

**Discussion by rpenaran** [microsoft/msquic#2384] [Q&A] [answered]
Hi, I've just started using this library un Ubuntu systems. I've been testing the throughput obtained using secnetperf in 2 machines (let's say A and B) with the same configuration and HW. When I run the test with download option (client ask for data to the server) I achieve what I'm suppose to achieve (~7-8 Gbps) in both options, A as a server and B as a client and vice versa. However, when I use the same test but using the upload option (client sending data to the server), they are not reaching even 1 Gbps, in any configuration (A as a server and B as a client or vice versa). I saw that when downloading the core where the worker is suppose to be assigned is at 100% of utilization but when uploading, that is not happening (barely 25%). I also read the code of the test and the code to send data from the server and from the client is very similar. 

Do you experience similar behavior? Do you have any clue why this is happening?

Thanks!
Roberto

**Comment by nibanks** [comment, marked as answer]
Our performance tests don't seem to reproduce what you describe: https://microsoft.github.io/msquic/

<img width="819" alt="image" src="https://user-images.githubusercontent.com/20663557/153481082-5bea5c70-d947-43b2-862f-a42a5cfd5366.png">

We do have https://github.com/microsoft/msquic/issues/1612 which is tracking the fact that Linux upload perf is bimodal. Maybe you're only hitting the "bad" case?

**Reply by rpenaran** [reply]
Maybe, for some reason I'm hitting the bad case, although it's even worse that your bad case :(
I see that this is an issue from a year ago. Has any progress been made?

**Reply by nibanks** [reply]
We haven't investigated it any. It's been one of those "it's good enough, and no one complains, so leave it for now" kind of issues. So, you're saying in your experience, no CPUs are running at 100% on the sender or receiver side? If so, there's something else obviously blocking throughput. The events that log [QUIC_FLOW_BLOCKED_REASON](https://github.com/microsoft/msquic/blob/main/src/inc/quic_trace.h#L44) should be able to tell you what's going on.

**Reply by rpenaran** [reply]
Thanks! I've been looking at the logs and I didn't see anything weird. I've been comparing both tests (download and upload) to see if there are differences different events. The events per packet transmission are the same in both cases and I didn't see any "block" event between them. This is an example for upload case:
![image](https://user-images.githubusercontent.com/72943635/153685733-5036fd0c-7982-4cdc-8ce1-b568d19601f6.png)

The only thing I found is that sometimes the time between events are very different. But I don't know the reason. I'll try to go deeper and see.
